dandi/download.py

```diff
@@ -197,112 +341,441 @@ def download(
+      A backend (girder or api) specific fixture for downloading some file into
+      path. It should be a generator yielding downloaded blocks.
+    size: int, optional
+      Target size if known
+    digests: dict, optional
+      possible checksums or other digests provided for the file. Only one
+      will be used to verify download
+    """
+    if op.lexists(path):
+        block = f"File {path!r} already exists"
+        if existing == "error":
+            raise FileExistsError(block)
+        elif existing == "skip":
+            yield _skip_file("already exists")
+            return
+        elif existing == "overwrite":
+            pass
+        elif existing == "refresh":
+            if mtime is None:
+                lgr.warning(
+                    f"{path!r} - no mtime or ctime in the record, redownloading"
+                )
+            else:
+                stat = os.stat(op.realpath(path))
+                same = []
+                if is_same_time(stat.st_mtime, mtime):
+                    same.append("mtime")
+                if size is not None and stat.st_size == size:
+                    same.append("size")
+                # TODO: use digests if available? or if e.g. size is identical
+                # but mtime is different
+                if same == ["mtime", "size"]:
+                    # TODO: add recording and handling of .nwb object_id
+                    yield _skip_file("same time and size")
+                    return
+                lgr.debug(f"{path!r} - same attributes: {same}.  Redownloading")
+
+    if size is not None:
+        yield {"size": size}
+
+    destdir = op.dirname(path)
+    os.makedirs(destdir, exist_ok=True)
+
+    yield {"status": "downloading"}
+
+    algo, digester, digest, downloaded_digest = None, None, None, None
+    if digests:
+        # choose first available for now.
+        # TODO: reuse that sorting based on speed
+        for algo, digest in digests.items():
+            digester = getattr(hashlib, algo, None)
+            if digester:
+                break
+        if not digester:
+            lgr.warning("Found no digests in hashlib for any of %s", str(digests))
+
+    # TODO: how do we discover the total size????
+    # TODO: do not do it in-place, but rather into some "hidden" file
+    for attempt in range(3):
+        try:
+            downloaded = 0
+            if digester:
+                downloaded_digest = digester()  # start empty
+            warned = False
+            # I wonder if we could make writing async with downloader
+            with open(path, "wb") as writer:
+                for block in downloader():
+                    if digester:
+                        downloaded_digest.update(block)
+                    downloaded += len(block)
+                    # TODO: yield progress etc
+                    msg = {"done": downloaded}
+                    if size:
+                        if downloaded > size and not warned:
```

Comment:
`warned` never gets set to `True`.  Should that be added here?

Reply:
yeah, set it to True in 5abd195384356f48e37729ed364f2b89e37cd210 ... hopefully we never run into this part of code.